JEP 18137: Scope of JAX NumPy & SciPy Wrappers (#18137)
jax-ml/jax: docs/jep/18137-numpy-scipy-scope.md

# JEP 18137: Scope of JAX NumPy & SciPy Wrappers

*Jake VanderPlas*

*October 2023*

Until now, the intended scope of {mod}`jax.numpy` and {mod}`jax.scipy` has been relatively
ill-defined. This document proposes a well-defined scope for these packages to better guide
and evaluate future contributions, and to motivate the removal of some out-of-scope code.

## Background
From the beginning, JAX has aimed to provide a NumPy-like API for executing code in XLA,
and a big part of the project’s development has been building out the {mod}`jax.numpy` and

> {mod}`jax.scipy` namespaces as XLA-backed wrappers of NumPy and SciPy APIs. There has always

Reviewer: ```suggestion
{mod}`jax.scipy` namespaces as JAX-based implementations of NumPy and SciPy APIs. There has always
```

... since they support all jax transformations.

### Axis 2: Array API Alignment
The second axis we consider focuses on the
[Python Array API Standard](https://data-apis.org/array-api/2022.12/): this is in some
senses a community-driven outline of which array operations are central to array-oriented
programming across a wide range of user communities. If an API in  {mod}`numpy` or {mod}`scipy` is
listed within the Array API standard, it is a strong signal that JAX should include it.
Using the example from above, the Array API standard includes several variants of
 {func}`numpy.unique` (`unique_all`, `unique_counts`, `unique_inverse`, `unique_values`) which
suggests that, despite the function not being precisely aligned with XLA, it is important
enough to the Python user community that JAX should perhaps implement it.

### Axis 3: Existence of Downstream Implementations
For functionality that does not align with Axis 1 or 2, an important consideration for
inclusion into JAX is whether there exist well-supported downstream packages that supply
the functionality in question. A good example of this is {mod}`scipy.optimize`: while JAX does
include a minimal set of wrappers of {mod}`scipy.optimize` functionality, a much more complete

> treatment exists in the [JAXOpt](https://jaxopt.github.io/) package, which is actively

Reviewer: ```suggestion
treatment exists in the [JAXopt](https://jaxopt.github.io/) package, which is actively
```

to `dm-pix` or to another appropriate package.

#### ❌ `scipy.odr`
The {mod}`scipy.odr` module provides an object-oriented wrapper around `ODRPACK` for
performing orthogonal distance regressions. It is not clear that this could be cleanly
expressed using existing JAX primitives, and so we deem it out of scope for JAX itself.

#### ❌ `scipy.optimize`
The {mod}`scipy.optimize` module provides high-level and low-level interfaces for optimization.
Such functionality is important to a lot of JAX users, and very early on JAX created
{mod}`jax.scipy.optimize` wrappers. However, developers of these routines soon realized that
the {mod}`scipy.optimize` API was too constraining, and different teams began working on the
[JAXopt](https://jaxopt.github.io/) package and the
[Optimistix](https://github.com/patrick-kidger/optimistix) package, each of which contain
a much more comprehensive and better-tested set of optimization routines in JAX.

Because of these well-supported external packages, we now consider {mod}`scipy.optimize`
to be out-of-scope for JAX.

*Recommendation: deprecate {mod}`jax.scipy.optimize` and/or make it a lightweight wrapper

> around an optional JAXOpt or Optimistix dependency.*

Reviewer: ```suggestion
around an optional JAXopt or Optimistix dependency.*
```
